Create Service Class

garrincha33 opened the pull request:
Latest Chages

Removed the Protocol for now in the DataFetchService until we actually need it and removed all the spacing as requested. When your happy to approve will merge to main

garrincha33 committed:
added codemagic yaml file for develop branch

garrincha33 committed:
removed storyboard from project and added orange background to ViewController

garrincha33 committed:
created the mainTabBarController with 2 tabs and added icons to tabBar

garrincha33 committed:
updated indentation, created 2 new functions for tabColor scheme and insets, created enum for sizes, removed all comments

garrincha33 committed:
added APIKey secret holding plist added to gitignore file, also created A constant, Service and Endpoints file. Tested that I could call the FULL_CRYPTO api with apikey imbedded

ananogal commented on CryptoLightNews/Constants/BaseConstant.swift:
```diff
@@ -0,0 +1,34 @@
+        self.dict = plist
+    }
+}
+class Constants: APIBase, APIKeyable {
```
Do you need a `class` here? Or with a `struct` is sufficient? For what I understand, this is just for holding constants... Do you even need a `struct` or an `enum` will "do the job"?

ananogal commented on CryptoLightNews/Networking/Service.swift:
```diff
@@ -0,0 +1,53 @@
+
+import Foundation
+
+class Serivce {
```
Typo
```suggestion
class Service {
```

ananogal commented on CryptoLightNews/Networking/Service.swift:
```diff
@@ -0,0 +1,53 @@
+import Foundation
+
+class Serivce {
+    static let shared = Serivce()
```
Why are we creating a Singleton here? Do we really need a Singleton?

ananogal commented on CryptoLightNews/Networking/Service.swift:
```diff
@@ -0,0 +1,53 @@
+        fetchGenericJSONData(urlString: urlString.rawValue, completion: completion)
+    }
+    
+    func fetchGenericJSONData<T: Decodable>(urlString: String, completion: @escaping (Result<T, CryptoNewsErrors>) -> ()) {
```
Maybe you should verify the URL before you call this function. And make it private.  
```suggestion
    private func fetchGenericJSONData<T: Decodable>(url: URL, completion: @escaping (Result<T, CryptoNewsErrors>) -> ()) {
```

garrincha33 replied:
struggling with this one unable to get it to work with the suggestion

ananogal commented on CryptoLightNews/Networking/Service.swift:
```diff
@@ -0,0 +1,53 @@
+    }
+    
+    func fetchGenericJSONData<T: Decodable>(urlString: String, completion: @escaping (Result<T, CryptoNewsErrors>) -> ()) {
+        guard let url = URL(string: urlString) else {
+            completion(.failure(CryptoNewsErrors.networkError))
+            return
+        }
```
Put it in the `fetchNews`
```suggestion
       
```

garrincha33 replied:
struggling with this one unable to get it to work with the suggestion

ananogal commented on CryptoLightNews/Networking/Service.swift:
```diff
@@ -0,0 +1,53 @@
+            }
+            guard let httpReponse = response as? HTTPURLResponse, (200...299).contains(httpReponse.statusCode) else {
+                print("no data from response")
+                completion(.failure(CryptoNewsErrors.networkError))
```
Maybe we should start to think about creating a different kinds of errors: if you always return the same one, when do you know if your error was a parsing error or a network error or a URL typo error? Give each error a name. It will help you to debug in the future. Print is a debugging tool; only use it to debug and never push it to production.

garrincha33 replied:
how do i show the error from the ifLet (currently within print) removing print(error) obviously gives me the yellow warning that im not using the unwrapped value

ananogal commented on CryptoLightNews/Model/NewsArticles.swift:
```diff
@@ -0,0 +1,27 @@
+    var urlToImage: String?
+    var author: String?
+    
+    enum CodingKeys: String, CodingKey {
```
Do you need to create the CodingKeys here? It the keys are the same as your properties' names, you don't need them.

ananogal commented on CryptoLightNews/Networking/Endpoints.swift:
```diff
@@ -0,0 +1,32 @@
+            return URL(string: NEW_FULL_COINS)!
+        }
+    }
+    var urlRequest: URLRequest {
```
Do we need this var? Because in `Service`, you are not using it, and you are verifying the URL in `fetchNews`. 
Be careful with what you add to your code. Have a look at this code principle [YAGNI](https://betterprogramming.pub/yagni-you-aint-gonna-need-it-f9a178cd8e1)

ananogal commented on .gitignore:
```diff
@@ -117,4 +117,8 @@ fastlane/test_output
 
 iOSInjectionProject/
 
+.DS_Store
```
This is not working since I see the file is pushed to GitHub... make sure you have the correct path in the .gitignore file.

garrincha33 committed:
Updating Create ServiceClass

garrincha33 committed:
add more cases for Error Ahndling

garrincha33 committed:
Final Changes for Create Service Class

ananogal commented on CryptoLightNews/Networking/DataFetchService.swift:
```diff
@@ -0,0 +1,28 @@
+        self.networkService = networkService
+    }
+}
+
```
Please remove the extra spaces

ananogal commented on CryptoLightNews/Networking/DataFetchService.swift:
```diff
@@ -0,0 +1,28 @@
+
+import Foundation
+
+protocol Fetchable {
```
Where are you using this protocol? If not in use, please delete it.

ananogal reviewed (changes requested):
Very good! just left a small comment.

garrincha33 committed:
Changes for Create Service Class

garrincha33 commented:
Latest Chages

Removed the Protocol for now in the DataFetchService until we actually need it and removed all the spacing as requested. When your happy to approve will merge to main

garrincha33 committed:
updating creating service class

garrincha33 commented:
added .plist to the file, this looks correct to me, when the file is being ignored i will get a new key

garrincha33 committed:
fixing gitignore

garrincha33 committed:
trying to fix gitignore
